### PR TITLE
locking for Misbehave() and other cs_main locking fixes

### PR DESCRIPTION
**Pull request by kazcw**
In several places in main.cpp, synchronized data structures are accessed without cs_main being held.

### CODE REVIEW

**Comment by gmaxwell** [comment]
hmph. peppering cs_main in the middle of functions sounds like a formula for lock inversion.  Perhaps Misbehaving() should get its own lock internally and not depend on cs_main?

**Comment by kazcw** [comment]
A separate nodestate cs is definitely doable; there's a fair amount of code that would need to hold cs_main and nodestate, but no new locks ever need to be acquired within the scope of a cs_nodestate lock. I'll draft a refactor.

**Comment by kazcw** [comment]
I created a separate mutex, cs_misbehavior. cs_misbehavior is only locked either directly inside a cs_main lock or along with no other locks, so there shouldn't be any deadlocks (and DEBUG_LOCKORDER isn't complaining). cs_misbehavior is the sole mutex guarding the nMisbehavior and fShouldBan fields; modifications to the mapBlockSource container are guarded by both cs_main and cs_misbehavior so that either lock suffices to perform lookups safely.

File: src/main.cpp [hunk 1]
```diff
@@ -2948,7 +2950,7 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
                 {
                     LOCK(cs_vNodes);
                     BOOST_FOREACH(CNode* pnode, vNodes) {
-                        if (chainActive.Height() > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate)) {
+                        if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate)) {
```
Comment (sipa) [inline src/main.cpp]: You can use pindexNewTip->nHeight instead of creating an extra variable.

**Comment by sipa** [comment]
@kazcw I don't see your cs_misbehaviour, but I think the code here is fine as is. We certainly need a separate lock for node states, but I would delay that until the net refactors are done.

**Comment by sipa** [comment]
utACK 719de56ab2c8e5bc6ce9f67c7bf159adc242d49b

**Comment by dcousens** [comment]
utACK 719de56,  and I concur that I do not see `cs_misbehavior`

File: src/main.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/main.cpp]: > You can use pindexNewTip->nHeight instead of creating an extra variable.

ping @kazcw

**Comment by laanwj** [comment]
utACK 719de56
Going to merge this, the nit can be done later.

**Pull request merged**